DjangoProject/bookmarksapp/tests.py

```diff
@@ -0,0 +1,107 @@
+from django.test import TestCase
+from django.shortcuts import reverse
+
+from authenticationapp.models import MyUser
+from .models import Folders, Bookmarks
+
+
+class BookmarksAppTestCase(TestCase):
+    user = None
+
+    def setUp(self):
+        """
+        This function will execute before for each testcase.
+        User creation and login will happen.
+        Some sample bookmarks and folders will be save in database.
+        """
+        self.credentials = {
+            'username': 'pavan@gmail.com',
+            'password': 'pavankumar'
+        }
+        MyUser.objects.create_user(**self.credentials)
+        self.user = self.client.post(reverse('authenticationapp:login'), self.credentials, follow=True).context['user']
+        f = Folders.objects.create(
+            name="Google",
+            created_by=self.user
+        )
+        y = Folders.objects.create(
+            name="Yahoo",
+            created_by=self.user
+        )
+
+        Bookmarks.objects.create(
+            folder=f,
+            url="https://www.google.com",
+            name='Google Website',
+            created_by=self.user
+        )
+
+        Bookmarks.objects.create(
+            folder=y,
+            url="https://www.yahoo.com",
+            name='Yahoo Website',
+            created_by=self.user
+        )
+
+    def tearDown(self):
+        """
+        Logout the user after finishing every testcase.
+        """
+        self.client.get(reverse('authenticationapp:logout'))
+
+    def test_folder_list_without_login(self):
+        """
+        Test folderlist page without logged in the user.
+        """
+        self.client.get(reverse('authenticationapp:logout'))
+        response = self.client.get(reverse('bookmarksapp:index'))
+        self.assertRedirects(response, reverse('authenticationapp:login'))
+
+    def test_post_login_page(self):
+        """
+        Test to Login Page by posting credentials.
+        """
+        self.client.get(reverse('authenticationapp:logout'))
+        response = self.client.post(reverse('authenticationapp:login'), self.credentials)
+        self.assertRedirects(response, reverse('authenticationapp:index'))
+
+    def test_get_folder_list_view(self):
+        """
+        Test to get folder list page.
+        """
+        response = self.client.get(reverse('bookmarksapp:folders'))
+        folders = Folders.objects.filter(created_by=self.user)
+        self.assertQuerysetEqual(response.context['folders'], folders, transform=lambda x: x, ordered=False)
```

Comment:
- Test if the right template is used
- Test if correct context variable is passed to template
- Check if pages count is matching expected (Create 30 folder in loop)